### PR TITLE
8319875: Add macOS implementation for jcmd System.map

### PR DESCRIPTION
**Pull request by stooke**
This is a port of #16301  to macOS.

System.map and System.dump_map are implemented using the macOS API and provide roughly the same information in the same format. Most of the heavy lifting was implemented by @tstuefe in https://github.com/openjdk/jdk/pull/16301 - this PR adds the macOS implementation and enables the common code for macOS 64 bit.

The System.map tests are also reworked to be cleaner for the three implementations.

[sample output using G1GC](https://github.com/user-attachments/files/17558064/g1_map.txt)
[sample output using ZGC](https://github.com/user-attachments/files/17558067/zgc_map.txt)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319875](https://bugs.openjdk.org/browse/JDK-8319875): Add macOS implementation for jcmd System.map (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Sonia Zaldana Calles](https://openjdk.org/census#szaldana) (@SoniaZaldana - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20953/head:pull/20953` \
`$ git checkout pull/20953`

Update a local copy of the PR: \
`$ git checkout pull/20953` \
`$ git pull https://git.openjdk.org/jdk.git pull/20953/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20953`

View PR using the GUI difftool: \
`$ git pr show -t 20953`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20953.diff">https://git.openjdk.org/jdk/pull/20953.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20953#issuecomment-2434463149)
</details>

### CODE REVIEW

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:120]: You could reduce the code and remove duplications with an [x macro](https://en.wikipedia.org/wiki/X_macro) if you wanted.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:360]: I would make this an assert, and return without printing anything in release.

**Review by tstuefe — COMMENTED** [review]
Nice work, thank you. Small nits inside. I have no time to test this extensively, but if the tests run through, it should be fine.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:122]: Doing this, I found a copy-paste error, so thanks for the suggestion!

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp]: I missed this.  I have changed it to the Windows version, where it prints the error and has an assert(), instead of fatal().

File: test/hotspot/jtreg/serviceability/dcmd/vm/SystemMapTestBase.java
Comment (tstuefe) [inline test/hotspot/jtreg/serviceability/dcmd/vm/SystemMapTestBase.java:180]: ?? remnant

**Review by tstuefe — COMMENTED** [review]
Looking good, small nits remain. Could you share an example output (maybe one run with G1, one with ZGC?)

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:336]: Deleted line 335

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:79]: Good idea.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:196]: done.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:101]: gone.

File: test/hotspot/jtreg/serviceability/dcmd/vm/SystemMapTestBase.java
Comment (stooke) [reply test/hotspot/jtreg/serviceability/dcmd/vm/SystemMapTestBase.java:180]: Gone.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:212]: Done. My thinking was to separate the information gathering from the display code.

**Comment by stooke** [comment]
> Looking good, small nits remain. Could you share an example output (maybe one run with G1, one with ZGC?)

Sample output for G1 and ZGC has been attached to the description (and I fixed the ZGC test)

**Comment by tstuefe** [comment]
Hi @stooke,

nice. There are some aspects about the output that is surprising, though. I expect reserved-only memory to be mapped with MAP_NORESERVE and no access rights. I expect committed memory to be mapped without MAP_NORESERVE and write- or executable access.

- (G1) I only see one segment for JAVAHEAP. I would expect at least two, one reserved-only, one reserved and committed. Or did you precommit the heap (start JVM with -Xmx==-Xms)? Can you please test that with ParallelGC and without specifying Xms, we should see four java heap mappings side by side, for young and old gen, each with a committed part (rw) and a reserved part (not accessible). Like we do now for the CODE section. Please also countercheck that the address ranges are correct against the heap range (easiest, run jcmd .. VM.info and look for the Heap section, it tells you the address ranges).
- The 64K META and CLASS blocks confuse me. For one, they are too small. A Metaspace node is 64M sized. It can contain multiple committed and reserved sections, so typically spreads over multiple lines, but usually not many. Here (G1 file), I see many that are 64K in size, spanning in total just about 5MB. You can counter-check the META blocks with `jcmd .. VM.metaspace vslist` - it prints the virtual space node list at the start.
- Another thing that is strange is that the MEAT lines are not coalesced, though the properties displayed seem to be the same. But that can be the result of hidden properties being different.

Also, could you rename the OS-specific names to `[lowercase-something]`? The `[UPPERCASE]` names are easily confused with the uppercase NMT tags.

Thanks!

**Comment by stooke** [comment]
Hello, @tstuefe , and thanks for your comments.  I'll address a few here while I work on the others.
I have changed the os-specific names to lowercase, but I don't think it makes them stand out more.  The square brackets were intended to do that.  Might I change this back?

I think there is only one JAVAHEAP segment because due to an issue with my build[1] there was no CDS archive available.

I will look at the META and CLASS entries and see if there are hidden properties that I can surface, or if there's another reason for so many entries.

[1] I had given a target architecture to the configure command, which turned on cross-compiles (which disables CDS archive building) even when building on the target platform.

**Comment by tstuefe** [comment]
Hi @stooke !

> Hello, @tstuefe , and thanks for your comments. I'll address a few here while I work on the others. I have changed the os-specific names to lowercase, but I don't think it makes them stand out more. The square brackets were intended to do that. Might I change this back?

Sure, if it looks worse. I just wanted to make sure we can cleanly distinguish NMT sections from OS sections.

> 
> I think there is only one JAVAHEAP segment because due to an issue with my build[1] there was no CDS archive available.

Has nothing to do with CDS. The heap consists of committed and reserved areas. Committed areas have backing swap space allocated for them, and are accessible. Reserved areas have not and are generally not. API wise the difference is that Reserved sections set the MAP_NORESERVE flag for mmap, and are generally allocated with PROT_NONE. 

So, the heap should show up with several neighboring sections, some committed, some just reserved. Similar how most of the stacks should show up with two entries, one for the writable stack, one for the guard page that is protected.

---


Simple test: I reserve 1G of memory at startup, uncommitted (added to os::init_2)

```
  if (UseNewCode) {
    char* p = os::reserve_memory(G, false, mtInternal);
    tty->print_cr("Pointer is %p", p);
  }
```


vmmap shows:

```
VM_ALLOCATE                 10ccb4000-14ccb4000    [  1.0G     0K     0K     0K] ---/rwx SM=NUL  
```

so, looks good. 1GB, with all protection flags cleared. But System.map shows nothing for this address range.


Now, I commit the second half of the range:


```
  if (UseNewCode) {
    char* p = os::reserve_memory(G, false, mtInternal);
    tty->print_cr("Pointer is %p", p);
    bool b = os::commit_memory(p + (512 * M), 512 * M, false);
    assert(b,"???");
  }
```

vmmap shows only the committed part now, omitting the still uncommitted first half. But it gets the protection flags right again (rw now):

```
VM_ALLOCATE (reserved)      148000000-168000000    [512.0M     0K     0K     0K] rw-/rwx SM=NUL          reserved VM address space (unallocated)
```

System.map shows nothing.


What goes on? Is the OS lying to us? Do we have an error? Both vmmap and System.map seem to struggle, with vmmap being somewhat more correct.

**Comment by stooke** [comment]
@tstuefe I ran an experiment with raw mmap, and there's no way to differentiate between one large allocation of 5*128MB and 5 smaller allocations of 128MB.  I _could_ add code to fold these, but we risk loosing information.

**I have found that the combination of mach_make_memory_entry_64() + mach_vm_map() (instead of using mmap(), which is just a compatibility layer on top of mach system calls) causes the kernel to treat the allocation as one region, and report it as such!  This may be a topic of a further PR**

**Comment by tstuefe** [comment]
> @tstuefe I ran an experiment with raw mmap, and there's no way to differentiate between one large allocation of 5*128MB and 5 smaller allocations of 128MB. I _could_ add code to fold these, but we risk loosing information.

What information would we loose?

As it is now, the display is somewhat confusing. We did not allocate the heap with multiple mmap calls, each one of 128MB in size; we use a single mmap call.

If you want to close the work for now and leave this glitch for later, we can do this too.

**Comment by stooke** [comment]
@tstuefe if it's up to me, I would leave the folding for a quick later PR (in fact I would start it right after this one goes in).
I also would like to investigate the use of mach_make_memory_entry_64() which could be interesting on it's own.

Do you know how I can get the GitHub runner to start working?  It seems one of them is misconfigured.

**Comment by tstuefe** [comment]
> @tstuefe if it's up to me, I would leave the folding for a quick later PR (in fact I would start it right after this one goes in). I also would like to investigate the use of mach_make_memory_entry_64() which could be interesting on it's own.
> 
> Do you know how I can get the GitHub runner to start working? It seems one of them is misconfigured.

No idea, but it was broken for a while now, wasn't it? If you figure it out and fix it a lot of ppl would be thankful :)

Otherwise, since the code does not touch anything dangerous, I think its fine to check it in as long as the other platforms are green and you have executed the relevant test on MacOS for System.map and System.dump_map

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:107]: remnant?

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:99]: this makes me nervous. Please do a bounds check beforehand. Assert is fine. Who knows if Apple changes the numeric definitions for SM_xxx

Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:117]: Done.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp]: Done.

**Comment by stooke** [comment]
I've actually added the combining code here, and reconfirmed that both the System.map and System.dump_map tests pass on my arm M2.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:105]: We use // in cpp for most cases

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:121]:
```suggestion
    assert(SM_COW == 1 && SM_LARGE_PAGE == (sizeof(share_strings)/sizeof(share_strings[0]), "share_mode contants are out of range"); // the +1 offset is intentional, see below
```

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:105]: done.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:121]: done.

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (tstuefe) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:49]: Can you fix the whitespace error, please? jcheck should pass.

**Review by tstuefe — APPROVED** [review]
Okay!

Needs a second review. @ashu-mehra, maybe?

**Comment by tstuefe** [comment]
BTW make sure you merge and test again before integrating. Your branch is quite old. In fact, you should merge upstream often (The rule with merge vs rebase is: you merge after the patch has been un-drafted und RFR mail has been set out; before that, you can rebase).

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (SoniaZaldana) [inline src/hotspot/os/bsd/memMapPrinter_macosx.cpp:352]: Should we use memset instead?

**Review by SoniaZaldana — APPROVED** [review]
I am not a Reviewer but this looks good to me. I left a small comment inline. Thanks Simon!

**Review by tstuefe — APPROVED** [review]
Still good. Ship it!

File: src/hotspot/os/bsd/memMapPrinter_macosx.cpp
Comment (stooke) [reply src/hotspot/os/bsd/memMapPrinter_macosx.cpp:352]: A very good question!  Changing it now would invalidate the reviews, unfortunately.  Portability is not a concern here as it's for macOS only.  Since bzero only ever zeros memory, it _can_ be slightly faster: [one discussion](https://fdiv.net/2009/01/14/memset-vs-bzero-ultimate-showdown).

**Pull request closed without merging**
